ctru-rs/src/services/ir_user.rs

```diff
@@ -0,0 +1,491 @@
+                    packet_info[2],
+                    packet_info[3],
+                ]) as usize;
+                let data_length = u32::from_ne_bytes([
+                    packet_info[4],
+                    packet_info[5],
+                    packet_info[6],
+                    packet_info[7],
+                ]) as usize;
+
+                // Find the packet data. The packet data may wrap around the buffer end, so
+                // `packet_data` is a function from packet byte offset to value.
+                let packet_info_section_size = user_state.recv_packet_count * PACKET_INFO_SIZE;
+                let header_size = SHARED_MEM_RECV_BUFFER_OFFSET + packet_info_section_size;
+                let data_buffer_size = user_state.recv_buffer_size - packet_info_section_size;
+                let packet_data = |idx| -> u8 {
+                    let data_buffer_offset = offset_to_data_buffer + idx;
+                    shared_mem[header_size + data_buffer_offset % data_buffer_size]
+                };
+
+                // Find out how long the payload is (payload length is variable-length encoded)
+                let (payload_length, payload_offset) = if packet_data(2) & 0x40 != 0 {
+                    // Big payload
+                    (
+                        ((packet_data(2) as usize & 0x3F) << 8) + packet_data(3) as usize,
+                        4,
+                    )
+                } else {
+                    // Small payload
+                    ((packet_data(2) & 0x3F) as usize, 3)
+                };
+
+                // Check our payload length math against what the packet info contains
+                assert_eq!(data_length, payload_offset + payload_length + 1);
+
+                // IR packets start with a magic number, so double check it
+                let magic_number = packet_data(0);
+                assert_eq!(magic_number, 0xA5);
+
+                IrUserPacket {
+                    magic_number: packet_data(0),
+                    destination_network_id: packet_data(1),
+                    payload_length,
+                    payload: (payload_offset..payload_offset + payload_length)
+                        .map(packet_data)
+                        .collect(),
+                    checksum: packet_data(payload_offset + payload_length),
+                }
+            })
+            .collect()
+    }
+
+    /// Internal helper for calling ir:USER service methods.
+    fn send_service_request(
+        &self,
+        mut request: Vec<u32>,
+        expected_response_len: usize,
+    ) -> crate::Result<Vec<u32>> {
+        let mut shared_mem_guard = IR_USER_STATE.lock().unwrap();
+        let shared_mem = shared_mem_guard.as_mut().unwrap();
+
+        // Set up the request
+        let cmd_buffer = unsafe {
+            &mut *(slice_from_raw_parts_mut(
+                ctru_sys::getThreadCommandBuffer(),
+                max(request.len(), expected_response_len),
+            ))
```

Comment:
Is it sound to create an `&mut[T]` from a pointer like this (since theoretically other parts of the application could modify it while the `&mut` is held)?

Example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=bab65a45e045afd8c3f22bc9d5322149
Both `y` and `z` claim unique `&mut` access to the slice at the same time which I think is technically undefined behavior (works in this case but not always).

I guess we might make similar assumptions in other places sometimes but I'm not positive that we should... `ptr::copy` on the other hand I think should normally work fine.

Reply:
We shouldn't. This goes against Rust's safety standards. Moreover, we should wrap these methods with a proper (even if just internal) ′svc′ wrapper.